### PR TITLE
Fix: Navigation arrows not navigating along 'CalendarList'

### PR DESCRIPTION
**Pull request by olfek**
Replaces #631 

Before fix:
before_fix

After fix:
after_fix

These gifs can be explained by [this comment:](https://github.com/wix/react-native-calendars/pull/631#discussion_r279191482)

> Also, If I remember correctly, this is **needed** for navigating months in `CalendarList`.
> 
> This is because `CalendarList` renders a list of `CalendarListItem` which renders a `Calendar` which includes a `CalendarHeader`, when the left/right navigation arrows of the `CalendarHeader` are pressed, instead of navigating along the `CalendarList`, the `Calendar` is updated in place to show the month navigated to.
> 
> So for example, if `CalendarList` looked like:
> 
> ```
> -------+-------+-------+-------+
>  Index |   0   |   1   |   2   |
> -------+-------|-------|-------|
>  Month |  Jan  |  Feb  |  Mar  |
> -------+-------+-------+-------+
> ```
> 
> When `Jan` is shown, if the right arrow is pressed, instead of moving to the next `CalendarItem` in the `CalendarList` at `1`, The `CalendarItem` at `0` updates itself to show `Feb`. So now, `CalendarList` looks like:
> 
> ```
> -------+-------+-------+-------+
>  Index |   0   |   1   |   2   |
> -------+-------|-------|-------|
>  Month |  Feb  |  Feb  |  Mar  |
> -------+-------+-------+-------+
> ```
> 
> This is happening because in `CalendarHeader`:
> 
> ```
> onPressLeft() {
>     const {onPressArrowLeft} = this.props;
>     if(typeof onPressArrowLeft === 'function') {
>         return onPressArrowLeft(this.substractMonth);
>     }
>     return this.substractMonth();
> }
> 
> onPressRight() {
>     const {onPressArrowRight} = this.props;
>     if(typeof onPressArrowRight === 'function') {
>         return onPressArrowRight(this.addMonth);
>     }
>     return this.addMonth();
> }
> ```
> 
> which uses:
> 
> ```
> addMonth() {
>     this.props.addMonth(1);
> }
> 
> substractMonth() {
>     this.props.addMonth(-1);
> }
> ```
> 
> which uses `addMonth` in `Calendar`:
> 
> ```
> addMonth(count) {
>     this.updateMonth(this.state.currentMonth.clone().addMonths(count, true));
> }
> ```
> 
> which **replaces** the month displayed by `Calendar`.
> 
> This behaviour is incorrect when `Calendar` is used **inside** a `CalenderList`. When the left/right navigation arrows of `CalendarHeader` are pressed, we need to **move along** the `CalendarList` and display the previous/next `CalendarItem`, not update the month of the displayed `Calendar` item held by `CalendarList`.
> 
> To fix this, I added the following code to `CalendarList` which uses the changes in this MR:
> 
> ```
> onPressArrowLeft={(_,month)=>{
>     month = month.clone()
>     month.addMonths(-1)
>     this.refs.calendarList.scrollToMonth(month)
> }}
> onPressArrowRight={(_,month)=>{
>     month = month.clone()
>     month.addMonths(1)
>     this.refs.calendarList.scrollToMonth(month)
> }}
> ```
> 
> Where `calendarList` is a reference to the `CalendarList`.

@Inbal-Tish New MR

### CODE REVIEW

**Comment by olfek** [comment]
@Inbal-Tish ...

**Comment by olfek** [comment]
@Inbal-Tish Tests have now passed

**Comment by olfek** [comment]
@Inbal-Tish 

The actions in the GIFs above, are as follows.

**Before fix**

1. Swipe left to go from May to June. Observe that it says May and then June.
1. Swipe right to go back from June to May.
1. Click the right navigation arrow. Observe that it now says June.
1. Swipe left to go to the next month. Observe that the next month is also June.
1. Swipe right to go back to previous month.
1. Click the left navigation arrow until it says April.
1. Swipe right to go to the preceding month. Observe that this also says April.
1. Swipe left to go back to the month we were at in the last step.
1. Click right navigation arrow to make April go back to May.
1. Swipe right again to go to the preceding month. Observe that things are now correct again.

**After fix**

1. Swipe left to go from May to June. Observe that it says May and then June.
1. Swipe right to go back from June to May.
1. Click the right navigation arrow. Observe that is now says June.
1. Swipe left to go the next month. Observe that the next month is now July.
1. Swipe right to go back to the previous month.
1. Click the left navigation arrow until it says April.
1. Swipe right to go the preceding month. Observe that this is now March.
1. Swipe left to go back to the month we were at in the last step.
1. Click right navigation arrow to make April go back to May.
1. Swipe right again to go to the preceding month. Observe that that this is now April.

**Comment by olfek** [comment]
@Inbal-Tish Please can you merge this, I have been waiting 8 months!

File: src/calendar-list/index.js [hunk 1]
```diff
@@ -187,7 +187,8 @@ class CalendarList extends Component {
 
   renderCalendar({item}) {
     return (
-      <CalendarListItem 
+      <CalendarListItem
+        calendarListContainer={this}
```
Comment (Inbal-Tish) [inline src/calendar-list/index.js]: Passing this/ref of the list to every item is not a good practice. You should use a callback which you then invoke in the onPress functions. For example: 
moveMonth = (month) => { 
    this.scrollToMonth(month); 
} 
and pass it to the item:  moveMonth={this.moveMonth}
and invoke it safely from the item: _.invoke(this.props, 'moveMonth');

File: src/calendar-list/item.js [hunk 2]
```diff
@@ -47,8 +47,16 @@ class CalendarListItem extends Component {
           disabledByDefault={this.props.disabledByDefault}
           showWeekNumbers={this.props.showWeekNumbers}
           renderArrow={this.props.renderArrow}
-          onPressArrowLeft={this.props.onPressArrowLeft}
-          onPressArrowRight={this.props.onPressArrowRight}
+          onPressArrowLeft={this.props.onPressArrowLeft ? this.props.onPressArrowLeft : (_,month)=>{
```
Comment (Inbal-Tish) [inline src/calendar-list/item.js]: Using arrow functions in props is not advisable, you should move the functionality to separate methods. 
Also, you should check if 'this.props.calendarListContainer' is not undefined before invoking its method (if you use a callback, as I suggested in the first comment, you can instead call it safely using Lodash's 'invoke'). 
Third, you shouldn't assign to an argument, declare a new const for 'month' clone.

File: src/calendar-list/index.js (hunk 1 above)
Comment (olfek) [reply src/calendar-list/index.js]: Ok, I will see what I can do

**Comment by olfek** [comment]
@Inbal-Tish What do you think?

**Comment by olfek** [comment]
Tested it too, everything still works :+1:

**Comment by olfek** [comment]
@Inbal-Tish Resolved merge conflicts, tested again, still works. Please merge.

File: src/calendar-list/item.js [hunk 3]
```diff
@@ -47,13 +71,13 @@ class CalendarListItem extends Component {
           disabledByDefault={this.props.disabledByDefault}
           showWeekNumbers={this.props.showWeekNumbers}
           renderArrow={this.props.renderArrow}
-          onPressArrowLeft={this.props.onPressArrowLeft}
-          onPressArrowRight={this.props.onPressArrowRight}
+          onPressArrowLeft={this.onPressArrowLeft.bind(this)}
```
Comment (Inbal-Tish) [inline src/calendar-list/item.js]: Please don't use bind in props. You can use the arrow function on the methods' signatures. For example:
`onPressArrowRight = (_, month) => {
    ...
}`

Comment (olfek) [reply src/calendar-list/item.js]: @Inbal-Tish I've made the requested changes. Please merge.

File: src/calendar-list/item.js [hunk 4]
```diff
@@ -47,13 +71,13 @@ class CalendarListItem extends Component {
           disabledByDefault={this.props.disabledByDefault}
           showWeekNumbers={this.props.showWeekNumbers}
           renderArrow={this.props.renderArrow}
-          onPressArrowLeft={this.props.onPressArrowLeft}
-          onPressArrowRight={this.props.onPressArrowRight}
+          onPressArrowLeft={this.onPressArrowLeft}
```
Comment (Inbal-Tish) [inline src/calendar-list/item.js]: I would limit this fix to horizontal only, so we won't break the current behaviour of CalendatList in vertical for those who choose to add the navigation arrows for some reason. Consider something like:
`onPressArrowLeft={this.props.horizontal ? this.onPressArrowLeft : this.props.onPressArrowLeft}
  onPressArrowRight={this.props.horizontal ? this.onPressArrowRight : this.props.onPressArrowRight}`

Comment (olfek) [reply src/calendar-list/item.js]: @Inbal-Tish Okay, I've made these changes.

**Comment by olfek** [comment]
@Inbal-Tish merge?

**Pull request merged**